### ISSUE TITLE
Change log: 1.x.x to 2.x.x

### ISSUE THREAD

**Issue by bigearsenal** [p2p-org/solana-swift#42]
# Edited
- SolanaSDK namespace will be marked as deprecated, new classes, modules will be implemented without namespace.
- Better logger has to be integrated

# The problems and their solutions

### Problem

Due to lack of knowledges and documentation about Solana blockchain at the beginning of the development, the network client module, blockchain client module and other module are highly coupled in one class `SolanaSDK`, which is inconvenient for using and really hard to test.

For example: For retrieving list of supported token or calling `getBalance` for an account, the class `SolanaSDK` needs to be initialized first. As a result, all another modules has to be initialized too, including `AccountStorage`, which requires user to authenticate, even the SolanaRPC does not require authorization.

The one-class-to-rule-them-all is a bad idea, as it make the class unscalable and hard to test. The graph below shows the implementation of `SolanaSDK` at the moment:

<img width="1029" alt="SolanaSDK__(class)" src="https://user-images.githubusercontent.com/6975538/163972716-78d78089-1c93-4c68-8c29-1b386f14330f.png">

### Solution

The big class `SolanaSDK` has to be separated into multiple modules.

The term “independent” in this situation means:

- The module can be initialized and used independently. For example:`APIClient` must needs only endpoint to be initialized and can be used independently without authorization.
- The module has their own responsibility and doesn't need to know about another modules.

All modules has to be written in POP, which always starts with an interface (protocol). For example: `class APIClient: SolanaAPIClient`.

The `SolanaSDK` will be marked as deprecated for backward compatible.

The graph of `SolanaSDK` after refactoring is something like this:

<img width="1771" alt="SolanaSDK_-_after" src="https://user-images.githubusercontent.com/6975538/163972782-b81cf3fd-4b9a-4123-a081-4df13e45cc0e.png">

Modules’ explaination:

1. `AccountStorage: SolanaAccountStorage` stores user’s credentials and define if user is authorized.
2. `APIClient: SolanaAPIClient` api client to connect with Solana RPC API, contains following methods ([https://docs.solana.com/developing/clients/jsonrpc-api](https://docs.solana.com/developing/clients/jsonrpc-api)).
3. `TokensRepository: SolanaTokensRepository` repository that provides informations about supported token in Solana blockchain. ([https://raw.githubusercontent.com/solana-labs/token-list/main/src/tokens/solana.tokenlist.json](https://raw.githubusercontent.com/solana-labs/token-list/main/src/tokens/solana.tokenlist.json))
4. `BlockchainClient: SolanaBlockchainClient` the client that helps us preparing, signing and serializing transaction using `SolanaProgram`s before sending using method `sendTransaction`, `simulateTransaction` from `SolanaAPIClient` or forwarding to `FeeRelayer`.
5. `SolanaProgram` Type of Programs that lives on Solana Blockchain that provides methods for preparing instructions in any transaction.
6. `TransactionParser: SolanaTransactionParser`: Custom parser to define readable type of transaction, for example: Send, Swap, Receive,...
7. `Socket: SolanaSocket`: Socket connection that provide suitable methods to observe user’s account or transaction signature status.

### Problem

The SolanaSwift library is right now deeply depending on RxSwift.

The SolanaSwift library is also depending on RxAlamofire for handling network request, it must be replaced by the native solution to reduce dependency.

<img width="870" alt="Untitled" src="https://user-images.githubusercontent.com/6975538/163972980-40b7fa4a-bdb2-41da-9136-9c1e5539c389.png">

### Solution

As `swift concurrency` is now backward compatible with iOS 13, we use `async/await` and `URLSession` as the replacement for `RxSwift` and `RxAlamofire`

### Notice

In the scope of this refactoring, there would be no ui part touched, so a bridge needs to be created to convert `async/await` code to adapt new concurrency solution in UI (the UI will be refactored later). 

### Problem

Model is coupled inside `SolanaSDK`

### Solution

- Move models outside of `SolanaSDK`, for example `SolanaSDK.PublicKey` becomes `PublicKey`
- For backward compatibility, type alias has to be added back to `SolanaSDK`. For example:
```swift
extension SolanaSDK {
   typealias PublicKey = SolanaSwift.PublicKey // (?)
}
```

# **What is “neсessary”, and what is “nice to have”?**

1. The `SolanaSDK` is separated into multiple independent modules, *which are located in separated folder*. Еach module has to be fully tested.
2. The `SolanaSwift` needs to be independent from `RxSwift`, `RxAlamofire`.
3. The concurrency code is lightweight and easy to use.
4. A bridge need to be built to keep the ui part unchanged (will be refactored later)

# Task decomposition

This task has to be separated to child tasks, which equivalent to cleaning, deprecating and creating each modules in `SolanaSDK`:

1. Group all deprecated implementation of `SolanaSDK` in to `Deprecated` folder, move all Models out of `SolanaSDK`, but don't break the old working solution.
2. Implement `AccountStorage: SolanaAccountStorage`, blocked by 1
3. Implement `APIClient: SolanaAPIClient`, blocked by 1
4. Implement `TokensRepository: SolanaTokensRepository`, blocked by 1
5. Implement `BlockchainClient: SolanaBlockchainClient`, blocked by 1
6. Implement `SolanaProgram` , blocked by 1
7. Implement `TransactionParser: SolanaTransactionParser`, blocked by 1
8. Implement `Socket: SolanaSocket`, blocked by 1

# Definition of done

- `RxSwift`, `RxAlamofire` is removed from SolanaSwift’s dependencies, a bridge is created to map async function to return Observable to work with current Rx code in p2p-wallet-ios
- Several modules in `SolanaSDK` are independent
- All modules are fully tested
